### PR TITLE
Prefer explicit MDBList episodes over season summaries

### LINKED ISSUE #661: [BUG] Completed status on partially seen shows
**Describe the bug**
Shows got Completed status despite they are not fully watched.
**To Reproduce**
Import from MDblist
**Expected behavior**
Their status should be "In progress"
**Screenshots**

Image

**Error logs**
 If you have any error logs, paste them here:
No errors
**Floppy version**: v26.8.6-128-g36f821bf |Commit: 36f821b
**Database**: SQLite (default) or PostgreSQL
Sqlite

### LINKED ISSUE #662: [BUG] Wrong watched status after Mdblist import
**Describe the bug**
Whole season is marked as watched(and completed) while I only watched 1 episode.

**To Reproduce**
Import from Mdblist
**Expected behavior**
Correct watch states
**Screenshots**

Image

Image

**Error logs**
 If you have any error logs, paste them here:

**Floppy version**: v26.8.6-128-g36f821bf |Commit: 36f821b
**Database**: SQLite (default) or PostgreSQL 
Sqlite

### PR DESCRIPTION
**Pull request by srow90**
## Summary

- Prefer explicit MDBList episode rows over matching broad show or season summaries.
- Collect all cursor- or offset-paginated sync data before applying evidence precedence.
- Derive season completion from the complete distinct episode set, including existing completed episodes in `new` mode.
- Rebuild TV and season parents safely when overwrite mode replaces episode history.

## Problem

MDBList can return broad show or season summaries together with more precise episode rows. Processing a broad row first could expand every episode or mark durable history completed. Offset pagination also allowed the broad row and its matching episode evidence to land on different pages, while treating the highest episode number as complete could mark a partial season and show completed.

## Solution

- Use one page-advance helper for grouped sync and watchlist retrieval.
- Follow `next_cursor` when present; otherwise advance `offset` by the received row count while `has_more` is true.
- Retain request parameters and stop on empty, repeated, or non-advancing pages.
- Normalize nested and top-level show identifiers before applying evidence precedence.
- Suppress broad show/season evidence only for the matching show and season represented by valid detailed rows.
- Track distinct watched episode coordinates, seeded from completed episode rows in `new` mode.
- Complete a season only when its metadata episode set is covered; never infer show completion from the final episode position.
- Preserve `Dropped` during roll-up and avoid overwrite references to parents scheduled for deletion.

## Test matrix requested in review

- [x] Matching season summary and explicit episode on the same page.
- [x] Matching rows split across offset pages.
- [x] Final episode only does not complete a season.
- [x] A different show with the same season number is not suppressed.
- [x] Another season of the same show is not suppressed.
- [x] Nested and top-level show identifier shapes behave the same.
- [x] A show summary cannot override more detailed evidence.
- [x] Overwrite mode replaces episode history without foreign-key references to rows scheduled for deletion.

Additional coverage confirms cursor parameter retention, repeated-cursor termination, watchlist offset pagination, existing-episode completion seeding, and preservation of `Dropped`.

## Validation

- `scripts/test.sh integrations.tests.imports.test_mdblist` — 27 tests passed.
- The same 27-test MDBList suite passed in a disposable Python 3.12 Linux container.
- `uv run --no-sync ruff check src` — passed.
- `uv run --no-sync ruff format --check src/integrations/imports/mdblist.py src/integrations/tests/imports/test_mdblist.py` — passed.
- `git diff --check` — passed.
- `docker buildx build --platform linux/amd64 ...` — passed for the production Dockerfile.
- Native arm64 production-image health, migrations, assets, API metadata, MCP package, and restart-on-persisted-volume smoke — passed.
- Hosted PR lint, all CodeQL analyses, the full Python 3.12 application workflow, production-image smoke, and downstream image build — passed for commit `f51d0a31`.
- No repository QA command or `/gstack-qa` executable is available in this contributor environment.

An additional disposable Linux fast-suite run executed 3,887 tests. The MDBList and affected application tests passed; its three failures were confined to the ad hoc container harness (missing `git` and two nested-container entrypoint signal tests). The hosted Ubuntu application workflow above is the authoritative full-suite gate.

## AI Assistance

Generated with OpenAI Codex (`gpt-5.6-sol`, Codex 5.6 SOL). The final diff was reviewed against the repository contribution guidance and every maintainer review comment before submission.

## Contract handoff

- Domain guide: not applicable; no domain vocabulary changed.
- OpenAPI artifact: not applicable; no API contract changed.
- Migrations: not applicable; no models or migrations changed.
- Screenshots: not applicable; backend importer behavior only.

## Existing data

This prevents future incorrect imports. It does not automatically rewrite already stored history. The overwrite regression now verifies safe episode-history replacement, but users should still back up their database before using overwrite as a repair action.

## Human review

- [x] Maintainer review received; requested changes implemented in `f51d0a31`.
- [ ] Re-review completed.

Fixes #661.
Fixes #662.

### CODE REVIEW

File: src/integrations/imports/mdblist.py [hunk 1]
```diff
@@ -289,10 +304,23 @@ def process_watched(self):
...
         data = self._get_grouped_sync_data("/sync/watched", "watched entries")
+        explicit_episode_seasons = set()
+        for entry in data.get("episodes", []):
+            if not isinstance(entry, dict):
+                continue
+            episode_data = entry.get("episode") or {}
+            if not isinstance(episode_data, dict):
+                continue
+            show_data = episode_data.get("show") or entry.get("show") or {}
+            explicit_episode_seasons.update(
```
Comment (ryan-winkler) [inline src/integrations/imports/mdblist.py:380]: This pre-scan is correct only when `data` contains every provider page. `_get_grouped_sync_data()` currently follows `next_cursor`, while the MDBList sync response uses `offset` and `has_more`. Please add offset pagination before relying on this set. A regression should place the season summary on page 1 and its explicit episode on page 2; the result must remain one episode and `In progress`.

File: src/integrations/tests/imports/test_mdblist.py [hunk 2]
```diff
@@ -215,6 +215,45 @@ def test_watched_season_expands_to_episodes(self):
         season = Season.objects.get(item__media_id="12345", user=self.user)
         self.assertEqual(season.status, Status.COMPLETED.value)
 
+    def test_explicit_episode_prevents_season_fanout(self):
```
Comment (ryan-winkler) [inline src/integrations/tests/imports/test_mdblist.py:246]: This is a useful first regression, but episode 1 avoids the existing completion shortcut. Please add a sibling case with only episode 2 of this two-episode season. `_update_completion_status()` currently treats the highest numbered episode as proof that all earlier episodes were watched, so that case still becomes `Completed`. Completion must use the distinct watched episode set.

**Review by ryan-winkler — CHANGES_REQUESTED** [review]
Thanks for isolating the season fan-out and adding a focused regression test. The direction is correct: explicit episode evidence must take priority over a broad season summary.

I reviewed both changed files and followed the affected path through sync pagination, identifier resolution, bulk persistence, and the TV/season status model. I also checked the two linked bug reports and the current MDBList response contract.

I am requesting changes because two paths can still produce the same false completion that this PR is intended to remove.

## 1. Collect every sync page before applying precedence

`process_watched()` assumes that `_get_grouped_sync_data()` has collected the complete watched response before it builds `explicit_episode_seasons`. The current page loop only follows `pagination.next_cursor`.

MDBList documents the `/sync/*` response with `offset`, `limit`, and `has_more`. A user with more than one page can therefore get this sequence:

1. Page 1 contains a season summary.
2. A later page contains the matching explicit episode.
3. The importer stops after page 1.
4. The season summary expands into every TMDB episode.

That leaves the original bug in place for larger histories.

Please centralize page advancement and support both forms safely:

- follow `next_cursor` when a response supplies it;
- otherwise, when `has_more` is true, advance `offset` by the number of rows received;
- retain `limit` and other request parameters;
- stop when the response is empty or the next page parameters do not advance;
- use the same helper for grouped sync data and the watchlist.

Required regression: put the season summary on page 1 and the matching episode on page 2. The result must contain one watched episode, with the season and show still `In progress`.

## 2. Completion must require complete episode evidence

`_update_completion_status()` currently marks the season complete when the imported episode number equals `season_metadata["max_progress"]`. It then marks the show complete when that season is the last season.

This means a user who watched only the final episode of a season can still receive a completed season and completed show. The new test uses episode 1 of 2, so it does not exercise this path.

Please track distinct watched episode coordinates for each show and season. Include existing completed episode rows in `new` mode. Mark a season complete only when the watched episode set covers the complete episode set from the season metadata. Mark a show complete only when complete show evidence exists. Do not replace a `Dropped` status during this roll-up.

Required regression: import only episode 2 from a two-episode season. The episode must exist, but the season and show must remain `In progress`.

## 3. Define the show-summary boundary

The importer processes `shows` before `seasons` and `episodes`, and a show row is recorded as `Completed`. Please add one boundary test in which the same show has a broad show row and one explicit episode row. If MDBList can return both levels together, the detailed row must prevent or reconcile the broad completion. If the provider contract guarantees that these shapes cannot coexist, record that evidence in the PR.

## Test matrix before merge

Please add or confirm these cases:

- matching season summary and explicit episode on the same page;
- matching rows split across offset pages;
- final episode only does not complete a season;
- a different show with the same season number is not suppressed;
- another season of the same show is not suppressed;
- nested and top-level show identifier shapes behave the same;
- a show summary cannot override more detailed evidence;
- overwrite mode replaces episode history without leaving foreign-key references to rows scheduled for deletion.

The last case is important before we recommend overwrite mode as a repair step for affected users.

## Current QA evidence

The current head has passing application tests, lint, Docker build, and CodeQL checks. Those checks confirm that the branch is stable, but they do not cover the boundary cases above. After the update, please run the focused MDBList suite, the full application test workflow, Ruff check, Ruff format check, and the repository QA command.

## Security and user experience review

I did not find a credential leak, authorization bypass, injection path, or code-execution issue in this diff. The material risk is durable user-history corruption and incomplete imports from untrusted provider data.

This is a backend-only change. Screenshots are correctly marked as not applicable. No visual, keyboard, screen-reader, or cognitive-load behavior changes in this PR. Keep any new warning text short, specific, and actionable.

I could not attach a maintainer commit because the connected GitHub app cannot write to the contributor fork, even though maintainer edits are enabled. The inline comments and this review give the implementation and test boundaries without taking ownership away from the contributor. Please update this thread after the next revision so the data-integrity cases can be rechecked.

File: src/integrations/imports/mdblist.py (hunk 1 above)
Comment (srow90) [reply src/integrations/imports/mdblist.py:380]: Resolved in `f51d0a31`. `_next_page_params` is now shared by grouped sync and watchlist retrieval, supports cursor and offset/has_more contracts, retains parameters, and stops on empty or non-advancing pages. The offset regression splits the season summary and explicit episode across pages and remains one episode with season/show In progress.

File: src/integrations/tests/imports/test_mdblist.py (hunk 2 above)
Comment (srow90) [reply src/integrations/tests/imports/test_mdblist.py:246]: Resolved in `f51d0a31`. Completion now uses the distinct watched episode set, seeded from existing completed episode rows in new mode, and requires coverage of the metadata episode set. The explicit final-episode-only regression keeps both season and show In progress.
